### PR TITLE
apf: take seats into account when dispatching request

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
take into account `seats` when we dispatch a request:
- if the number of `seats` asked for by a request exceeds the concurrency limit of the priority level then that request can execute when the number of currently executing requests for the given priority level is zero
- if `seats in use + seats asked for by a request > concurrency limit` then we don't dispatch the request
 

#### Special notes for your reviewer:

This is work in progress, if you are okay with this quick fix, i can prepare it for review with tests added.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by tkashem** [comment]
@wojtek-t @MikeSpreitzer let me know if the idea of this quick fix is good enough for supporting `width`. 

/assign @MikeSpreitzer @wojtek-t

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 1]
```diff
@@ -245,7 +245,7 @@ func (qs *queueSet) StartRequest(ctx context.Context, width uint, hashValue uint
 	// Step 0:
 	// Apply only concurrency limit, if zero queues desired
 	if qs.qCfg.DesiredNumQueues < 1 {
-		if qs.totSeatsInUse >= qs.dCfg.ConcurrencyLimit {
+		if !qs.canAccomodateSeatsLocked(int(width)) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:248]: nit: instead of converting to int, let's just make the param of canAccomodateSeatsLocked uint :)

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 2]
```diff
@@ -662,6 +662,29 @@ func (qs *queueSet) cancelWait(req *request) {
 	qs.obsPair.RequestsWaiting.Add(-1)
 }
 
+func (qs *queueSet) canAccomodateSeatsLocked(seats int) bool {
+	switch {
+	case seats > qs.dCfg.ConcurrencyLimit:
+		// we have picked the queue with the minimum virtual finish time,
+		// but the number of seats this request asks for exceeds the concurrency limit
+		// TODO: this is a quick fix for now, once we have borrowing in place we will not need it
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:670]: What is described in the KEP is this concept of additional latency. Copying from the KEP:
```
One more important aspect to resolve is what happens if a given priority level doesn't have enough concurrency units assigned to it. To be on the safe side we should probably implement borrowing across priority levels. However, given we don't want to block introducing the width concept on design and implementation of borrowing, until this is done we have two main options:

cap the width at the concurrency units assigned to the priority level
reject requests for which we won't be able to allocate enough concurrency units
To avoid breaking some users, we will proceed with the first option (when computing the cap we should also report requests that we believe are too wide for a given priority level - it would allow operators to adjust configs). That said, to accommodate for the inaccuracy here we will introduce a concept of additional latency for a request. This basically means that after the request finishes in a real world, we still don't mark it as finished in the virtual world for additional latency. Adjusting virtual time of a queue to do that is trivial. The other thing to tweak is to ensure that the concurrency units will not get available for other requests for that time (because currently all actions are triggerred by starting or finishing some request). We will maintain that possibility by wrapping the handler into another one that will be sleeping for additional latence after the request is processed.

Note that given the estimation for duration of processing the requests is automatically corrected (both up and down), there is no need to change that in the initial version.
```

I don't necessary need this to be implemented in this PR, but let's just add a TODO to implement this (this isn't that hard to add and borrowing will take us a lot of time).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 3]
```diff
@@ -683,6 +706,17 @@ func (qs *queueSet) selectQueueLocked() *queue {
 			}
 		}
 	}
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:709]: I started wondering about picking the queue by "VirtualFinish" time. Up until now, it actually worked fine.

But now, it may become extremely penalizing for wider requests.
Imagine a case where there are just three queue:
- in first two queues you have a number of extremely fast requests with width=1 (constant incoming flow of them so that the queue always have one)
- in the last queue you have a request with say width=2 (but let's say it will also be really fast)
- say we have concurrency limit = 2
Now - because we have G=60s, what is happening is:
queue1: finish_time = t0 + G
queue2: finish_time = t0 + G
queue2: finish_time = to + 2G
[we pick up queue1 and queue2 to dispatch their requests]
let's say they finish after 1ms and we course-correct their virtual time and now we have:
queue1: finish_time = t0 + 1ms + G
queue2: finish_time = t0 + 1ms + G
queue3: finish_time = t0 + 2G
and now we again dispatch from queue1 and queue2

and so on

So until we dispatch something from queue3, it will take ages...


So I think we should actually choose the queue based on VirtualStartTime (as opposed to VirtualFinishTime) and just probably resolve conflicts by preferring narrower requests.

@MikeSpreitzer @tkashem - thoughts?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 4]
```diff
@@ -683,6 +706,17 @@ func (qs *queueSet) selectQueueLocked() *queue {
...
+	// TODO: add a method to fifo that lets us peek at the oldest request
+	var oldestReqFromMinQueue *request
+	minQueue.requests.Walk(func(r *request) bool {
+		oldestReqFromMinQueue = r
+		return false
+	})
+	if !qs.canAccomodateSeatsLocked(oldestReqFromMinQueue.Seats()) {
+		return nil
+	}
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:719]: I wrote a long comment why it doesn't work and then I realized it works because we don't modify robinIndex below (we don't modify if we don't dispatch anything).
Can you maybe add an explicit comment above that it's expected and we do that to preserve round-robin behavior?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 2 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:670]: BTW - I prototyped that (and couple other things) in: https://github.com/kubernetes/kubernetes/pull/102875
Feel free to pick up that PR too (though we should probably add additionallatency in a separate PR not in this one).

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:248]: please ignore

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 2 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:673]: thanks, I am happy to pick it up, yeah, i will do it in a separate PR.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 4 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:724]: yes, I added a comment, and also the test I added verifies expected value of robin index with multiple calls to `selectQueueLocked`

**Comment by tkashem** [comment]
@wojtek-t @MikeSpreitzer it's ready for another pass

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:724]: Thanks!

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:712]: This PR has to be rebased on top of your last PR (this will cause some of your new tests to start failing). 

But instead of changing tests - we should in my opinion address this comment.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:712]: FWIW - for requests with width=1, this would be a no-op change effectively

**Comment by tkashem** [comment]
@wojtek-t I see this failure in `pull-kubernetes-unit`
```
: TestMain 
FAIL	k8s.io/kubernetes/pkg/api/testing	180.354s
```

not sure if this is related to this PR!

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 5]
```diff
@@ -789,6 +789,180 @@ func TestTotalRequestsExecutingWithPanic(t *testing.T) {
...
+					),
+				},
+			},
+			beforeSelectQueueLocked: func(attempt int, qs *queueSet) {
+				if attempt == 3 {
+					qs.dCfg.ConcurrencyLimit = 50
+				}
+			},
+			attempts:              3,
+			minQueueIndexExpected: []int{-1, -1, 1},
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:916]: I claim this test is incorrect - in connection to the comment that I added in the main code.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 3 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:711]: so i tracked how `queue.virtualStart` changes in order of the following events:

A: a new request arrives (enqueue)
- if the queue is inactive (no request is currently executing or no other request is in queue) then new request can start immediately (current `R(t)` value):
```
queue.virtualStart = qs.virtualTime
```
- if the queue is active then we don't change `queue.virtualStart`

B: next round - select a queue with the minimum finish time
- we make sure per-queue virtual time `queue.virtualStart` does not fall behind the global

C: the request is dequeued, we adjust `queue.virtualStart` as below
```
queue.virtualStart += G*W
```

D: request finishes - we do course correcion
```
r.queue.virtualStart -= (G-S)*W
```

the scenario you described, yes it would be a while before q3 is picked and that's fair i think if we consider the bit by bit round robin, at some point virtual finish time for q3 will be minimum


 

If we consider the the above operations, it seems we do take into account the arrival time and implement
```
Si = MAX(R(t), Fi-1)
```

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:711]: At some point it *would* be minimal I agree. The problem is huge how G is (it's 60s).
So given we're multiplying it by `seats`, it may take minutes until it would happen. Up until then we wll be course-correcting virtualStart (point D) that will still be pushing other queues to have smaller time.

And because of that, at some point we will simply reject that calls (as it would be waiting for too old).

Which makes this unfair, because once request with high width reaches the peak of queue, we effectively stop choosing that queue for a long time of there are requests with smaller width in other queue (assuming the wide request didn't arrive minutes earlier).

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:711]: In other words - if we assume that processing each request takes 1ms (which isn't completely unrealistic), then you have to process 60,000 requests from queue1 and queue2 before you pick the request from queue3. Given that the second request has just 2x higher width, that's not fair at all.

I strongly believe that just using "virtual_start" is the right thing to do:
- for requests with width=1, it's a no-op (because finish_time = start_time + G so finish_time_1 < finish_time_2 <=> start_time_1 < start_time_2)
- it achieves what we want for requests with higher width

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:711]: so let's simplify the scenario you described:
- we have two queues `q1` and `q2`
- `q1` has an infinite supply of requests with `W=1`
- `q2` has one request waiting in the queue with the same width `W` = 1
- requests complete really fast `S=1ms` 

In this scenario we will execute 60,000 requests from `q1` before we pick the request from `q2`. since we are adjusting the virtual start for the queue it may look like we are being unfair to `q2` with a request of same width. Yes it gets proportionally worse for request with `W` > 1. 


So what you are suggesting would be equivalent to what we have today, not take into account `W` into virtual finish time calculation (basically `W` = 1 for all requests). For now, if we don't use `W` in calculating virtual finish time then we are retaining the current behavior. This feels safer, at least for now.
I think we need unit tests where we can simulate these use cases in a controlled manner and observer/assert on the expected behavior with multiple queues and requests with varying width. This would require some refactor of the `queueSet` and carefully crafted unit tests.

In the meantime, we could have the following:
- do not take `W` into account when calculating virtual finish time - in the virtual world all requests are estimated to take the same amount of time.
- we use `W` for seats allocation.

@wojtek-t @MikeSpreitzer thoughts?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:711]: > q2 has one request waiting in the queue with the same width W = 1

You mean q2 has request with w=2, right?

> - do not take W into account when calculating virtual finish time - in the virtual world all requests are estimated to take the same amount of time.
>  - we use W for seats allocation.

So to rephrase:
- we still use W to adjusting virtualStart as we do now, e.g. here:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L640
- but we just change the GetNextFinish() to not take `W` into account

That translates to exactly the same what I'm proposing (i.e. taking virtualStart, because virtualFinish for every request will be equal to "virtualStart + G". So I don't really see how this is different from what I'm proposing.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:711]: > You mean q2 has request with w=2, right?

no, q2 has request with w=1 but I forgot to mention concurrency limit = 1

> That translates to exactly the same what I'm proposing

I was referring to:
- A: when we dequeue: `queue.virtualStart += G`
- B: when request finishes: `queue.virtualStart -= G-S`
- C: when we pick: finish time `queue.virtualStart + G`

This is the current behavior, what you are suggesting is for `C` we don't use `W`, but otherwise use it for A and B. with your proposal, while there is a currently executing request with much higher width, smaller requests waiting in the said queue won't be picked up since its `virtualStart` is higher because we included `W` and these requests waiting have a chance to timeout as well. 
with my proposal we are equally fair to all requests, `W` matters in only seat allocation. it's practically no change from how we select from queue today. 

we can start with what you are proposing, we can tweak later if need be. @MikeSpreitzer unless you have any other concerns i want to go ahead and make the changes

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:711]: > no, q2 has request with w=1 but I forgot to mention concurrency limit = 1

With width=1 everything works correctly. Because virtualStart will be increasing after every request so it doesn't matter.

> This is the current behavior

Current where? The current code behavior is already that we increase virtualStart by G*Seats() - see link I pasted above:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L640

> with your proposal, while there is a currently executing request with much higher width, smaller requests waiting in the said queue won't be picked up since its virtualStart is higher because we included W and these requests waiting have a chance to timeout as well.

Which is exactly what I want. We need to be fair across queue - not fair across requests.
This has to be adressed by changing how we choose to which queue we add a request - see:
https://github.com/kubernetes/kubernetes/pull/102875/files#r653296345


> with my proposal we are equally fair to all requests, W matters in only seat allocation. it's practically no change from how we select from queue today.

But that's not what we want.
Imagine two queues:
- q1 we have a number of short 'get object X' requests (width=1)
- q2 we have a number of large 'get all pods" requests (e.g. width=10)

What we really want is we want to provide equal capacity across these queue. So if the request is e.g. 10x more expensive we want to proceed 10x less requests from that queue.

So to summarize - we want to be fair across queues. What you're proposing doesn't provide that guarantee - wide requests to some extent will be starving small short requests, which is not what we want.
What I'm proposing provide the fairness across queues.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 5 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:916]: this test does not exercise the entire enqueu, select, dequeue and finish path, it just ensures that the given a snapshot of a `queueSet`, `selectQueueLocked` returns the queue with the minimum virtual finish time and also there are enough seats available

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 6]
```diff
@@ -662,6 +662,32 @@ func (qs *queueSet) cancelWait(req *request) {
...
+// seats available to accommodate a request with the given number of seats,
+// otherwise it returns false.
+func (qs *queueSet) canAccommodateSeatsLocked(seats int) bool {
+	switch {
+	case seats > qs.dCfg.ConcurrencyLimit:
+		// we have picked the queue with the minimum virtual finish time, but
+		// the number of seats this request asks for exceeds the concurrency limit.
+		// TODO: this is a quick fix for now, once we have borrowing in place we will not need it
+		if qs.totRequestsExecuting == 0 {
+			return true
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:675]: Can you please add a TODO to set additional_latency once we add support to it?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go [hunk 7]
```diff
@@ -104,24 +104,19 @@ func (q *queue) Dequeue() (*request, bool) {
...
 
-	// the estimated service time of the oldest request is (G * request width)
-	estimatedServiceTime := float64(G) * float64(oldestReq.Seats())
-	return q.virtualStart + estimatedServiceTime
+	// the estimated service time of the oldest request is:
+	//   virtual start time + G
+	// We are not taking the width of the request into account when
+	// we calculate the virtual finish time of the request because
+	// it can starve requests with smaller wdith in other queues.
+	return q.virtualStart + float64(G)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go:119]: I think in the current form we have that inconsistent with other places where we actually modify virtualStart with "G x Seats()".

I would actually suggest changing this method to "GetNextStart() float64" and simply returning q.virtualStart.

Given that we only use this function in one place and only comparison matters (and we use common G everywhere), I it would actually be much more intuitive and consistent with all other places.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go]: actually i just removed it for now, since it's a one-liner.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 8]
```diff
@@ -674,14 +701,33 @@ func (qs *queueSet) selectQueueLocked() *queue {
 		qs.robinIndex = (qs.robinIndex + 1) % nq
 		queue := qs.queues[qs.robinIndex]
 		if queue.requests.Length() != 0 {
-			currentVirtualFinish := queue.GetNextFinish(qs.estimatedServiceTime)
+			// the virtual finish time of the oldest request is:
+			//   virtual start time + G
+			// We are not taking the width of the request into account when
+			// we calculate the virtual finish time of the request because
+			// it can starve requests with smaller wdith in other queues.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:708]: one last nit: can you maybe add a specific example that we were discussing (I mean your simplified version with two queues) and discussion about it (e.g. need to process 60,000 requests before we put from the other queue) to ensure we won't forget about this case in the future?

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:708]: done, pls check

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 9]
```diff
@@ -674,14 +701,44 @@ func (qs *queueSet) selectQueueLocked() *queue {
...
+			//
+			// so let's draw an example of the starving scenario:
+			//  - G=60 (estimated service time in seconds)
+			//  - concurrency limit=2
+			//  - we have two queues, q1 and q2
+			//  - q1 has an infinite supply of requests with width W=1
+			//  - q2 has one request waiting in the queue with width W=2
+			//  - virtual start time for both q1 and q2 are at t0
+			//  - requests complete really fast, S=1ms on q1
+			// in this scenario we will execute roughly 30,000 requests
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:718]: s/30,000/60,000/

[just being pedantic]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 10]
```diff
@@ -662,6 +662,33 @@ func (qs *queueSet) cancelWait(req *request) {
 	qs.obsPair.RequestsWaiting.Add(-1)
 }
 
+// canAccommodateSeatsLocked returns true if this queueSet has enough
+// seats available to accommodate a request with the given number of seats,
+// otherwise it returns false.
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:667]: I think this PR should update this comment, perhaps to something like the following.

> // canAccommodateSeatsLocked decides whether the QueueSet's concurrency constraint allows dispatching a new request that will occupy the given number of seats.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 8 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:708]: I do not think we should attempt a justification in a comment here.  The original fair queuing paper says that the dispatch order is the order of completion in the virtual world (with ties broken by round-robin ordering), and none of our design alternatives change that.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:708]: The example below shows that if we didn't change what we had before, it wouldn't work. So I actually claim that this comment is really needed.

I'm happy to hear how we can frame it differently, but I don't think that's very urgent.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go (hunk 10 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:667]: @tkashem - let's do that in a follow up

**Review by wojtek-t — COMMENTED** [review]
I'm going to approve to have this merged before leaving on vacation.
We can chat more about tuning comments once I'm back too (it seems Mike`s comments are only about the comments).
